Port some code to Python3

ghostd opened the pull request:
With this PR i wish to share modifications to run `mach` with `./mach` instead of `python3 ./mach`.

I'm using a Debian machine. The following commands work for me:
* `./mach build --dev`
* `./mach test-tidy`
* `./mach fmt`

I still have errors with the WPT scripts, but maybe the root causes is a bad resolution of `requirements.txt`

I used `2to3` to migrate some code before applying some manual fixes. For PLY, I copied/paste the code from the github repository.

If someone can test...
For now i'm not sure how to continue the migration while i can't run WPT :-/

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [ ] These changes fix #___ (GitHub issue number if applicable)

<!-- Either: -->
- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because ___

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

ghostd committed:
Port some code to Python3

highfive commented:
Heads up! This PR modifies the following files:
 * @asajeffrey: components/script/dom/bindings/codegen/CodegenRust.py, components/script/dom/bindings/codegen/ply/README, components/script/dom/bindings/codegen/ply/ply/lex.py, components/script/dom/bindings/codegen/run.py, components/script/build.rs and 4 more
 * @wafflespeanut: python/requirements.txt
 * @kichjang: components/script/dom/bindings/codegen/CodegenRust.py, components/script/dom/bindings/codegen/ply/README, components/script/dom/bindings/codegen/ply/ply/lex.py, components/script/dom/bindings/codegen/run.py, components/script/build.rs and 4 more

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify script code, but no tests are modified. Please consider adding a test!

ghostd commented:
I updated and rebased the PR. My [last issue](https://github.com/web-platform-tests/wpt/issues/27025) needs a [WPT fix](https://github.com/web-platform-tests/wpt/pull/26976) (and is specific to Python 3.9)

I can run bootstrap, build, test-wpt, fmt and test-tidy.

jdm commented:
```
Running the WPT lint on ./tests/wpt/web-platform-tests/...
{"path": "import-maps/data-driven/tools/format_json.py", "lineno": 15, "rule": "PARSE-FAILED", "message": "Unable to parse file"}
```
This is an interesting failure.

jdm commented:
That presumably comes from https://github.com/web-platform-tests/wpt/blob/98049d97b75b4edbc71ee857b2b424bfec82524d/tools/lint/lint.py#L687-L696.

jdm commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 5a7da274c0e73ff08db9b829da6f8f2ce2a9ca6e with merge c8b57e63c8e3ae99b2f30aaa30c26f16c16bb697...

jdm reviewed (approved):
The code changes look fine to me.

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/MoAfB2tATVGUMIYZ8abgkw)

jdm commented:
We should update https://github.com/servo/servo/blob/master/etc/taskcluster/decision_task.py to always use python3.

ghostd commented:
> ```
> Running the WPT lint on ./tests/wpt/web-platform-tests/...
> {"path": "import-maps/data-driven/tools/format_json.py", "lineno": 15, "rule": "PARSE-FAILED", "message": "Unable to parse file"}
> ```
> 
> This is an interesting failure.

I proposed a [PR](https://github.com/web-platform-tests/wpt/pull/27037) to fix it

ghostd commented:
> We should update https://github.com/servo/servo/blob/master/etc/taskcluster/decision_task.py to always use python3.

I added `python3` for each `mach` command and tried to add a `with_python3` method for the Windows build (but i have no clue if it works). This commit really needs a careful review :-)

jdm commented on etc/taskcluster/decisionlib.py:
```diff
@@ -637,6 +637,24 @@ def with_python2(self):
+            path="python3",
+        ) \
+        .with_early_script("""
+            python -m ensurepip
+            pip install virtualenv==20.2.1
```
I think these may need to be `python3` and `pip3` respectively.

ghostd replied:
Done

jdm commented on etc/taskcluster/decisionlib.py:
```diff
@@ -637,6 +637,24 @@ def with_python2(self):
+        .with_early_script("""
+            python -m ensurepip
+            pip install virtualenv==20.2.1
+        """)
```
I believe this also requires `.with_path_from_homedir("python3", "python3\\Scripts")`

ghostd replied:
Done

bors-servo commented:
:umbrella: The latest upstream changes (presumably #28094) made this pull request unmergeable. Please resolve the merge conflicts.

tomoverlund commented:
How did you get ./mach build --dev without changes to mozjs?

jdm commented:
Presumably it was still using python2 for the mozjs build.

tomoverlund commented:
> Presumably it was still using python2 for the mozjs build.

Oh, that makes sense. I'm trying to build in Debian unstable with no python2 in sight.

ghostd commented:
I rebased and squashed the commit. The WPT PR has been merged \o/

jdm commented:
I'm working on incorporating these changes into #28092.

jdm commented:
I'm going to attempt to wrestle CI into shape in #28141.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #28147) made this pull request unmergeable. Please resolve the merge conflicts.

jdm commented:
I'm moving ahead with merging this code as part of #28092, because my attempts to get Windows CI working in #28141 didn't work out. I'm going to go ahead and close this PR—thank you so much for doing this work! It was a big help.
